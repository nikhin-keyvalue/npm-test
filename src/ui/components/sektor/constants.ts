const SECTOR_COLOR = '#FFFFFF';
const CIRCLE_COLOR = '#4F4F4F';

export { SECTOR_COLOR, CIRCLE_COLOR };
