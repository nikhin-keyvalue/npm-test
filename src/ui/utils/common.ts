import { ContractInterface, ethers } from 'ethers';
import { Cluster } from '@solana/web3.js';

import { AddEthereumChainParameter } from 'types/wallet';
import { USDCAbi, USDTAbi } from 'constants/abi';
import {
  MUMBAI_CHAIN_PARAMS,
  TokenTypes,
  EthChains,
  POLYGON_CHAIN_PARAMS,
  SolChains,
  SolanaCluster,
  ANDROID_WALLETCONNECT_DEEPLINK,
  CurrencyTypes,
  WalletConnectWalletNames
} from 'constants/wallet';
import { OS, REDIRECT_URLS, Wallets, WALLET_ICONS } from 'constants/common';
import { AvailableNetworks, ConnectionInformation, WalletData, WalletDetails } from 'types';
import { MOBILE_DEVICE_REGEX } from 'constants/regex';
import { getWalletMeta } from 'ui/wallets/WalletConnect';
import { ASSETS_PATH } from 'config';

const parseChain = (chain: EthChains | number) =>
  typeof chain === 'number' ? (`0x${chain.toString(16)}` as EthChains) : chain;

const parseAddress = (address: string) =>
  `${address.slice(0, 2)}${address.slice(2, 6).toUpperCase()}...${address.slice(-4).toUpperCase()}`;

const parseDecimals = (decimals: ethers.BigNumber | number) =>
  typeof decimals === 'number' ? decimals : decimals.toNumber();

const getChainParams = (chain: EthChains) => {
  if (chain === EthChains.MUMBAI) return MUMBAI_CHAIN_PARAMS;
  if (chain === EthChains.POLYGON) return POLYGON_CHAIN_PARAMS;

  return {} as AddEthereumChainParameter;
};

const getContractAbi = (token: TokenTypes) => {
  const contractAbi: ContractInterface = token === TokenTypes.USDT ? USDTAbi : USDCAbi;

  return contractAbi;
};

const getClusterFromChain = (chain: SolChains): Cluster => {
  if (chain === SolChains.DEVNET) return SolanaCluster.DEVNET;

  return SolanaCluster.MAINNET;
};

const getWalletName = (walletName: Wallets) => {
  switch (walletName) {
    case Wallets.METAMASK:
      return 'metamask';
    case Wallets.PHANTOM:
      return 'phantom';
    default:
      return '';
  }
};

const handleWalletInstallRedirect = (wallet: Wallets) => {
  const url = REDIRECT_URLS[wallet];

  window.open(url);
};

const isMobileDevice = () => MOBILE_DEVICE_REGEX.test(navigator.userAgent);

const triggerAndroidWalletDeepLink = () => {
  window.open(ANDROID_WALLETCONNECT_DEEPLINK);
};

const getLocalStorageWalletConnect = () => window.localStorage.getItem('walletconnect');

const isWalletConnect = (wallet: WalletData) => wallet.name === Wallets.WALLETCONNECT;

const isAndroid = (deviceOS: string) => deviceOS === OS.ANDROID;

const isIOS = (deviceOS: string) => deviceOS === OS.IOS;

const isTrustWallet = (walletName: string) =>
  walletName === WalletConnectWalletNames.TRUST_ANDROID || walletName === WalletConnectWalletNames.TRUST_IOS;

//TODO: image need to be fetched from BE
const getSelectedWalletIcon = (selectedWallet: WalletData) => {
  const walletMeta = getWalletMeta();

  // TODO: fallback for all popular wallets
  if (isWalletConnect(selectedWallet))
    return (
      (walletMeta?.icons && walletMeta?.icons[0]) ||
      `${ASSETS_PATH}/public/images/${
        isTrustWallet(walletMeta?.name || '') ? 'trust-wallet' : 'wallet-connect'
      }-icon.svg`
    );

  return `${ASSETS_PATH}/public/images/${getWalletName(selectedWallet.name)}-icon.svg`;
};

const getSelectedWalletName = (selectedWallet: WalletData) =>
  isWalletConnect(selectedWallet) ? getWalletMeta()?.name : selectedWallet.name;

const getFallbackWalletIcon = (wallet: Wallets) => `${ASSETS_PATH}/public/images/${WALLET_ICONS[wallet]}.svg`;

// TODO: compare with function in bitcoin branch
const copyToClipBoard = (copyText: string) => {
  navigator.clipboard.writeText(copyText);
};

const getBTCConnectionInfo = (availableNetworks: AvailableNetworks[], paymentOption: WalletDetails) => {
  const BTCNetwork =
    availableNetworks.find((item) => item.network_code === CurrencyTypes.BTC)?.network_code || CurrencyTypes.BTC;
  const connectionInfo: ConnectionInformation = {
    network_code: BTCNetwork,
    payment_option_id: paymentOption[Wallets.BTC].id
  };

  return connectionInfo;
};

export {
  parseAddress,
  parseChain,
  parseDecimals,
  getChainParams,
  getContractAbi,
  getClusterFromChain,
  getWalletName,
  getFallbackWalletIcon,
  handleWalletInstallRedirect,
  isMobileDevice,
  triggerAndroidWalletDeepLink,
  getLocalStorageWalletConnect,
  isWalletConnect,
  isAndroid,
  isIOS,
  getSelectedWalletIcon,
  getSelectedWalletName,
  copyToClipBoard,
  getBTCConnectionInfo
};
