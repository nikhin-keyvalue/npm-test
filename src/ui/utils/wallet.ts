import { ethers } from 'ethers';

import { Wallets } from 'constants/common';
import { CoinData } from 'types';
import { CurrencyTypes, EthChains, SolChains, TokenTypes } from 'constants/wallet';
import { Balance, ChainTypes, CoinTypes } from 'types/wallet';
import { getEthBalance, getTokenBalance } from '../wallets/common';
import { getSolBalance, getSolTokenBalance } from '../wallets/Phantom';

const isSolChain = (chain: ChainTypes): chain is SolChains => chain === SolChains.DEVNET || chain === SolChains.MAINNET;

const isEthChain = (chain: ChainTypes): chain is EthChains =>
  chain === EthChains.MAINNET ||
  chain === EthChains.RINKEBY ||
  chain === EthChains.POLYGON ||
  chain === EthChains.MUMBAI;

const isEthereumBasedCoin = (coin: CoinTypes): coin is CurrencyTypes.ETHEREUM | CurrencyTypes.MATIC =>
  coin === CurrencyTypes.ETHEREUM || coin === CurrencyTypes.MATIC;

const isSolBasedCoin = (coin: CoinTypes): coin is CurrencyTypes.SOL => coin === CurrencyTypes.SOL;

const isToken = (coin: CoinTypes): coin is TokenTypes => coin === TokenTypes.USDC || coin === TokenTypes.USDT;

const getCurrencyFromChain = (chain: ChainTypes) => {
  if (chain === EthChains.MAINNET || chain === EthChains.RINKEBY) return CurrencyTypes.ETHEREUM;
  else if (chain === EthChains.POLYGON || chain === EthChains.MUMBAI) return CurrencyTypes.MATIC;
  else if (chain === SolChains.DEVNET || chain === SolChains.MAINNET) return CurrencyTypes.SOL;

  return '' as CurrencyTypes;
};

const getEthBalancesArray = async (
  address: string,
  tokenList: CoinData[],
  chain: EthChains,
  selectedWallet: Wallets
) => {
  const balance = await getEthBalance(address, selectedWallet);
  const mainBalance: Balance = {
    coin: getCurrencyFromChain(chain),
    balance: Number(ethers.utils.formatEther(balance))
  };

  const tokenBalances = await Promise.all(
    tokenList.map(async (item) => {
      if (item) {
        const tokenBalance = await getTokenBalance(
          address,
          item.currency_code as TokenTypes,
          chain,
          item.contract_address,
          selectedWallet
        );

        return {
          coin: item.currency_code,
          balance: tokenBalance
        } as Balance;
      }

      return {
        coin: null,
        balance: 0
      };
    })
  );

  return [mainBalance, ...tokenBalances];
};

const getSolBalancesArray = async (address: string, tokenList: CoinData[], chain: SolChains) => {
  const balance = await getSolBalance(address, chain);
  const mainBalance: Balance = {
    coin: getCurrencyFromChain(chain),
    balance
  };

  const tokenBalances = await Promise.all(
    tokenList.map(async (item) => {
      if (item) {
        const tokenBalance = await getSolTokenBalance(address, item.contract_address, chain);

        return {
          coin: item.currency_code,
          balance: tokenBalance
        } as Balance;
      }

      return {
        coin: null,
        balance: 0
      };
    })
  );

  return [mainBalance, ...tokenBalances];
};

const isValidBalance = (balance: Balance[]) => {
  if (balance.length) {
    // TODO: make function
    const coins: CoinTypes[] = [...Object.values(CurrencyTypes), ...Object.values(TokenTypes)];
    let valid = true;

    balance.forEach((item) => {
      if (!(item.coin && coins.includes(item.coin))) valid = false;

      return;
    });

    return valid;
  } else {
    return false;
  }
};

export {
  isSolChain,
  isEthChain,
  isEthereumBasedCoin,
  isSolBasedCoin,
  isToken,
  getCurrencyFromChain,
  getEthBalancesArray,
  getSolBalancesArray,
  isValidBalance
};
