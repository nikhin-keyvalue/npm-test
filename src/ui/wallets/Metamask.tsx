import { ethers } from 'ethers';

import { AddEthereumChainParameter } from 'types/wallet';
import { getContractAbi } from 'ui/utils/common';
import {
  EthChains,
  TokenTypes,
  JsonRPCMethods,
  MetamaskEvents,
  MetamaskErrorCode,
  JsonRPCErrorCode
} from 'constants/wallet';
import { getWeb3ProviderAndSigner } from './common';

const getMetaMaskProvider = () => {
  const walletExist: any = window?.ethereum;
  let provider: any;

  // To handle multiple extension providers
  if (walletExist?.providers) provider = walletExist?.providers.find((prov: { isMetaMask: any }) => prov.isMetaMask);
  else provider = walletExist;

  return provider;
};

const connectMetaMask = async () => {
  const provider = getMetaMaskProvider();

  try {
    if (!provider) return null;

    const address: string[] = await provider.request({
      method: JsonRPCMethods.REQUEST_ACCOUNTS
    });

    return {
      success: true,
      address: address[0],
      chain: provider.chainId as EthChains,
      error: null
    };
  } catch (err: any) {
    return {
      success: err.code === MetamaskErrorCode.REQUEST_REJECT,
      address: null,
      chain: null,
      error: err.code as MetamaskErrorCode
    };
  }
};

const getMetamaskAccount = async () => {
  const provider = getMetaMaskProvider();

  try {
    if (!provider) return null;

    const address: string[] = await provider.request({
      method: JsonRPCMethods.ETH_ACCOUNTS
    });

    return address[0];
  } catch (err: any) {
    return null;
  }
};

const addMetaMaskWalletListener = (
  accountChangeHandler: (...args: any) => void,
  chainChangeHandler: (...args: any) => void
) => {
  const provider = getMetaMaskProvider();

  if (provider !== undefined) {
    provider.on(MetamaskEvents.ACCOUNT_CHANGE, accountChangeHandler);
    provider.on(MetamaskEvents.CHAIN_CHANGE, chainChangeHandler);

    return () => {
      if (window.ethereum !== undefined) {
        provider.removeListener(MetamaskEvents.ACCOUNT_CHANGE, accountChangeHandler);
        provider.removeListener(MetamaskEvents.CHAIN_CHANGE, chainChangeHandler);
      }
    };
  }

  return () => null;
};

const sendEth = async (from: string, to: string, value = '0') => {
  const provider = getMetaMaskProvider();

  try {
    if (!provider) throw new Error('No crypto wallet found. Please install it.');

    const { signer } = getWeb3ProviderAndSigner(provider as ethers.providers.ExternalProvider);

    const resp = await signer.sendTransaction({
      from,
      to,
      value: ethers.utils.parseEther(value)
    });

    return {
      success: true,
      transactionResponse: resp
    };
  } catch (err: any) {
    console.log(err);

    return {
      success: err.code !== JsonRPCErrorCode.INTERNAL_ERROR && err.code !== MetamaskErrorCode.INSUFFICIENT_FUNDS,
      transactionResponse: null
    };
  }
};

const transferToken = async (to: string, value = '0', token: TokenTypes, contractAddress: string) => {
  const provider = getMetaMaskProvider();

  try {
    const { signer } = getWeb3ProviderAndSigner(provider as ethers.providers.ExternalProvider);

    const contractAbi = getContractAbi(token);
    const contract = new ethers.Contract(contractAddress, contractAbi, signer);

    const resp = await contract.transfer(to, value);

    return {
      success: true,
      transactionResponse: resp
    };
  } catch (e: any) {
    console.log(e);

    return {
      success: e.code === JsonRPCErrorCode.INTERNAL_ERROR || e.code === MetamaskErrorCode.REQUEST_REJECT,
      transactionResponse: null
    };
  }
};

const addChain = async (chainParams: AddEthereumChainParameter) => {
  const provider = getMetaMaskProvider();

  try {
    await provider.request({
      method: JsonRPCMethods.ADD_CHAIN,
      params: [chainParams]
    });

    return true;
  } catch (e) {
    console.log(e);

    return false;
  }
};

const switchChain = async (chainId: string) => {
  const provider = getMetaMaskProvider();

  try {
    await provider.request({
      method: JsonRPCMethods.SWITCH_CHAIN,
      params: [
        {
          chainId
        }
      ]
    });

    return true;
  } catch (e) {
    console.log(e);

    return false;
  }
};

export {
  getMetaMaskProvider,
  connectMetaMask,
  getMetamaskAccount,
  addMetaMaskWalletListener,
  sendEth,
  transferToken,
  addChain,
  switchChain
};
