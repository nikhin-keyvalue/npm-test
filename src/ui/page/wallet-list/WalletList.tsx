import Preact, { h } from 'preact';
import { useEffect } from 'preact/hooks';

import { getChainParams, getFallbackWalletIcon, getLocalStorageWalletConnect, parseChain } from 'ui/utils/common';
import { addChain, addMetaMaskWalletListener, connectMetaMask, switchChain } from 'ui/wallets/Metamask';
import { connectProvider, connectWalletConnect } from 'ui/wallets/WalletConnect';
import { connectPhantom } from 'ui/wallets/Phantom';
import { EthChains, MetamaskErrorCode } from 'constants/wallet';
import { ModalTypes, Wallets } from 'constants/common';
import { ConnectionInformation, WalletListProps } from 'types';
import { ASSETS_PATH } from 'config';
import { getDeviceOS } from 'zamp/payment/utils';
import { isEthChain } from 'ui/utils/wallet';
import { ConnectionResponse } from 'types/wallet';

import './style.scss';

const WalletList: Preact.FunctionComponent<WalletListProps> = ({
  setSelectedWallet,
  setAddress,
  setCurrentChain,
  selectedChainObject,
  walletList,
  setRemoveWalletListener,
  onWalletConnected,
  setModalOpen,
  rpcUrls
}) => {
  const disconnectHandler = () => {
    setAddress(null);
  };

  const accountChangeHandler = (accounts: string | any[]) => {
    setAddress(accounts.length > 0 ? (accounts[0] as string) : null);
  };

  const chainChangeHandler = (newChain: EthChains | number) => {
    setCurrentChain(parseChain(newChain));
  };

  const handleMetamaskConnect = async () => {
    const resp = await connectMetaMask();
    let isOnValidChain = true;

    if (resp)
      if (resp.address) {
        const selectedChain = selectedChainObject.network_chain_id;

        if (resp.chain !== selectedChain) {
          isOnValidChain = false;
          let addChainResp = true;
          let switchChainResp = true;

          if (selectedChain === EthChains.MUMBAI || selectedChain === EthChains.POLYGON)
            addChainResp = await addChain(getChainParams(selectedChain));
          switchChainResp = await switchChain(selectedChain);
          if (switchChainResp) {
            setCurrentChain(selectedChain);
            setAddress(resp.address);
          }
          isOnValidChain = addChainResp && switchChainResp;
        } else {
          setCurrentChain(resp.chain);
          setAddress(resp.address);
        }
        const removeWalletListener = addMetaMaskWalletListener(accountChangeHandler, chainChangeHandler);

        setRemoveWalletListener(removeWalletListener as () => void);
        setModalOpen(null);
      } else if (!resp.success) {
        if (resp.error === MetamaskErrorCode.EXISTING_REQUEST) setModalOpen(ModalTypes.METAMASK_PENDING_REQUEST);
        else setModalOpen(ModalTypes.GENERIC_ERROR);
      } else {
        setModalOpen(null);
      }
    else setModalOpen(ModalTypes.WALLET_INSTALL);

    return isOnValidChain ? resp : null;
  };

  const handlePhantomClick = async () => {
    const resp = await connectPhantom();

    if (resp)
      if (resp.address) {
        setAddress(resp.address);
        setModalOpen(null);

        return resp;
      } else if (!resp.success) {
        setModalOpen(ModalTypes.GENERIC_ERROR);
      } else {
        setModalOpen(null);
      }
    else setModalOpen(ModalTypes.WALLET_INSTALL);
  };

  const handleWalletConnectClick = async () => {
    // TODO: parse to EthChains
    const resp = await connectWalletConnect(
      selectedChainObject.network_chain_id as EthChains,
      getDeviceOS(),
      accountChangeHandler,
      chainChangeHandler,
      disconnectHandler,
      rpcUrls
    );

    if (resp.address)
      if (resp?.chain === selectedChainObject.network_chain_id) {
        if (resp.chain === selectedChainObject.network_chain_id) {
          setCurrentChain(resp.chain);
          setAddress(resp.address);
        }

        return resp;
      } else if (!resp.success) {
        // TODO: ignore or add error notifier
      } else {
        setModalOpen(ModalTypes.INVALID_CHAIN);
      }
  };

  const onWalletSelect = async (item: Wallets) => {
    setSelectedWallet(walletList[item]);

    let connectionResponse: ConnectionResponse;
    let data: ConnectionInformation;

    if (item !== Wallets.WALLETCONNECT && item !== Wallets.BINANCE) setModalOpen(ModalTypes.WALLET_CONNECT);

    switch (walletList[item].name) {
      case Wallets.METAMASK:
        connectionResponse = await handleMetamaskConnect();
        break;
      case Wallets.PHANTOM:
        connectionResponse = await handlePhantomClick();
        break;
      case Wallets.WALLETCONNECT:
        connectionResponse = await handleWalletConnectClick();
        break;
      case Wallets.BINANCE:
        data = {
          payment_option_id: walletList[item].id,
          network_code: selectedChainObject.network_code
        };
        onWalletConnected(data, walletList[item].name);
        break;
      default:
        return;
    }

    if (connectionResponse?.address) {
      data = {
        payment_option_id: walletList[item].id,
        source_account: connectionResponse.address,
        network_code: selectedChainObject.network_code
      };

      onWalletConnected(data);
    }
  };

  const handleWalletClick = async (item: Wallets) => {
    await onWalletSelect(item);
  };

  // To handle wallet side WC Disconnects, must initialize this method to clear local storage
  useEffect(() => {
    if (isEthChain(selectedChainObject.network_chain_id) && getLocalStorageWalletConnect() && rpcUrls)
      connectProvider(selectedChainObject.network_chain_id, rpcUrls);
  }, [rpcUrls]);

  return (
    <div>
      {/* TODO: tobe removed */}
      <div className='zamp-wallet-container' role='presentation' key='nikhin' onClick={() => null}>
        <div className='zamp-wallet-label-container'>
          <img src={`${ASSETS_PATH}/public/icons/generic-crypto.svg`} alt='' className='wallet-icons' />
          <div className='wallet-name'>Pay with Exchange</div>
        </div>
        <div className='arrow-container'>
          <img src={`${ASSETS_PATH}/public/icons/exchange-icons.svg`} alt='' height='15' style='margin-right: 10px;' />
          <img src={`${ASSETS_PATH}/public/images/right-arrow.svg`} alt='' height='11' />
        </div>
      </div>
      <div className='zamp-wallet-container' role='presentation' key='nik' onClick={() => null}>
        <div className='zamp-wallet-label-container'>
          <img src={`${ASSETS_PATH}/public/icons/pay-with-qr.svg`} alt='' className='wallet-icons' />
          <div className='wallet-name'>Pay with QR</div>
        </div>
        <div className='arrow-container'>
          <img src={`${ASSETS_PATH}/public/images/right-arrow.svg`} alt='' height='11' />
        </div>
      </div>
      {/* Till here */}

      {selectedChainObject?.eligible_payment_options?.map((item) => {
        return (
          <div className='zamp-wallet-container' role='presentation' key={item} onClick={() => handleWalletClick(item)}>
            <div className='zamp-wallet-label-container'>
              <img
                src={walletList[item]?.logo_url}
                alt='wallet-icon'
                className='wallet-icons'
                onError={(e) => {
                  e.currentTarget.onerror = null;
                  e.currentTarget.src = getFallbackWalletIcon(item);
                }}
              />
              <div className='wallet-name'>{walletList[item]?.name}</div>
            </div>
            <div className='arrow-container'>
              <img src={`${ASSETS_PATH}/public/images/right-arrow.svg`} alt='' height='11' />
            </div>
          </div>
        );
      })}
    </div>
  );
};

export default WalletList;
