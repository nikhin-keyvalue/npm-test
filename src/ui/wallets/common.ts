import WalletConnectProvider from '@walletconnect/web3-provider';
import { ethers } from 'ethers';

import { Wallets } from 'constants/common';
import { ContractFunctions, EthChains, TokenTypes } from 'constants/wallet';
import { getContractAbi, parseDecimals } from '../utils/common';
import { getMetaMaskProvider } from './Metamask';
import { getWalletConnectProvider } from './WalletConnect';

const getContractData = (contract: ethers.Contract, functionName: ContractFunctions, params: any[]) => {
  const iface = contract.interface;
  const data = iface.encodeFunctionData(functionName, params);

  return data;
};

const getWeb3ProviderAndSigner = (provider: ethers.providers.ExternalProvider) => {
  const web3provider = new ethers.providers.Web3Provider(provider);
  const signer = web3provider.getSigner();

  return {
    web3provider,
    signer
  };
};

const getTokenDecimals = async (contract: ethers.Contract) => {
  const decimalsResp: ethers.BigNumber = await contract.decimals();

  return parseDecimals(decimalsResp);
};

const getEthProvider = (wallet: Wallets) => {
  switch (wallet) {
    case Wallets.METAMASK:
      return getMetaMaskProvider();
    case Wallets.WALLETCONNECT:
      return getWalletConnectProvider();
    default:
      return null;
  }
};

const getEthBalance = async (address: string, wallet: Wallets) => {
  try {
    const provider: ethers.providers.ExternalProvider | WalletConnectProvider = getEthProvider(wallet);

    const { web3provider } = getWeb3ProviderAndSigner(provider);
    const balance = await web3provider.getBalance(address);

    return balance;
  } catch (err: any) {
    return 0;
  }
};

const getTokenBalance = async (
  address: string,
  token: TokenTypes,
  chain: EthChains,
  contractAddress: string,
  wallet: Wallets
) => {
  try {
    const provider: ethers.providers.ExternalProvider | WalletConnectProvider = getEthProvider(wallet);
    const { signer } = getWeb3ProviderAndSigner(provider);

    const contractAbi = getContractAbi(token);
    const contract = new ethers.Contract(contractAddress, contractAbi, signer);

    const balanceResp: ethers.BigNumber = await contract.balanceOf(address);
    const decimals = await getTokenDecimals(contract);
    // TODO: toNumber breaks when balance to large
    const balance = balanceResp.toNumber();

    return balance / Math.pow(10, decimals);
  } catch (err: any) {
    console.log(err);

    return 0;
  }
};

export { getContractData, getWeb3ProviderAndSigner, getTokenDecimals, getEthBalance, getTokenBalance };
