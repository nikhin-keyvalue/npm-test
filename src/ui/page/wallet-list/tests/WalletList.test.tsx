// TODO: To be removed
/* eslint-disable @typescript-eslint/no-unused-vars */
import { h } from 'preact';
import { configure, mount, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { Adapter } from 'enzyme-adapter-preact-pure';

import WalletList from 'ui/page/wallet-list/WalletList';
import * as metamask from 'ui/wallets/Metamask';
import { EthChains } from 'constants/wallet';
import { AvailableNetworks, WalletDetails } from 'types';
import { Wallets } from 'constants/common';

// TODO: Fix issues
configure({ adapter: new Adapter() });

const walletList = [
  {
    id: 'probo_po_9GthjB6rbcvwb7eP5xid_05_27',
    name: 'Metamask',
    type: 'wallet'
  }
] as any;
const walletDetails: WalletDetails = {
  Metamask: {
    id: '',
    logo_url: '',
    name: Wallets.METAMASK,
    type: ''
  },
  Phantom: {
    id: '',
    logo_url: '',
    name: Wallets.PHANTOM,
    type: ''
  },
  WalletConnect: {
    id: '',
    logo_url: '',
    name: Wallets.WALLETCONNECT,
    type: ''
  },
  Binance: {
    id: '',
    logo_url: '',
    name: Wallets.BINANCE,
    type: ''
  },
  QRCode: {
    id: '',
    logo_url: '',
    name: Wallets.BTC,
    type: ''
  }
};

const selectedChainObject: AvailableNetworks = {
  network_chain_id: EthChains.RINKEBY,
  network_code: 'ETH',
  network_display_name: 'ETHEREUM',
  network_logo_url: '',
  eligible_payment_options: [Wallets.METAMASK],
  rpc_url: ''
};

const rpcUrls = new Map([[0, '']]);

describe('WalletList', () => {
  const mockFn = jest.fn();

  it('render components', () => {
    const wrapper = mount(
      <WalletList
        setSelectedWallet={mockFn}
        setAddress={mockFn}
        selectedChainObject={selectedChainObject}
        setCurrentChain={mockFn}
        walletList={walletDetails}
        setRemoveWalletListener={mockFn}
        onWalletConnected={() => null}
        setModalOpen={mockFn}
        rpcUrls={rpcUrls}
      />
    );
    const tree = shallowToJson(wrapper);

    expect(tree).toMatchSnapshot();
  });

  it('handle metamask connect click on success', async () => {
    const onHandleWalletsPageMock = jest.fn();
    const setSelectedWalletMock = jest.fn();
    const setRemoveWalletListenerMock = jest.fn();
    const setAddressMock = jest.fn();
    const setChainMock = jest.fn();

    const wrapper = shallow(
      <WalletList
        setSelectedWallet={setSelectedWalletMock}
        setAddress={setAddressMock}
        setCurrentChain={setChainMock}
        selectedChainObject={selectedChainObject}
        walletList={walletDetails}
        setRemoveWalletListener={setRemoveWalletListenerMock}
        onWalletConnected={() => null}
        setModalOpen={mockFn}
        rpcUrls={rpcUrls}
      />
    );

    const mock = jest.spyOn(metamask, 'connectMetaMask');

    mock.mockImplementation(async () => {
      return {
        success: true,
        address: '0x12345',
        chain: EthChains.RINKEBY,
        balance: 0,
        error: null
      };
    });

    // await wrapper.find('div.zamp-wallet-container').at(1).simulate('click');
    // expect(onHandleWalletsPageMock).toHaveBeenCalled();
    // expect(setSelectedWalletMock).toHaveBeenCalled();
    // expect(setAddressMock).toHaveBeenCalled();
    // expect(setChainMock).toHaveBeenCalled();
    // expect(setRemoveWalletListenerMock).toHaveBeenCalled();
  });

  it('handle metamask connect click on fail', async () => {
    const onHandleWalletsPageMock = jest.fn();
    const mockFn = jest.fn();

    const wrapper = shallow(
      <WalletList
        setSelectedWallet={mockFn}
        setAddress={mockFn}
        setCurrentChain={mockFn}
        selectedChainObject={selectedChainObject}
        walletList={walletDetails}
        setRemoveWalletListener={mockFn}
        onWalletConnected={() => null}
        setModalOpen={mockFn}
        rpcUrls={rpcUrls}
      />
    );

    const mock = jest.spyOn(metamask, 'connectMetaMask');

    mock.mockImplementation(async () => {
      return {
        success: false,
        address: null,
        chain: null,
        balance: 0,
        error: 4001
      };
    });

    // await wrapper.find('div.zamp-wallet-container').at(1).simulate('click');
    // expect(onHandleWalletsPageMock).toHaveBeenCalled();
  });
});
