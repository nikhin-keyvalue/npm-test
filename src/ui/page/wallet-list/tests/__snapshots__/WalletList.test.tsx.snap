// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletList render components 1`] = `
<WalletList
  onWalletConnected={[Function]}
  rpcUrls={
    Map {
      0 => "",
    }
  }
  selectedChainObject={
    Object {
      "eligible_payment_options": Array [
        "Metamask",
      ],
      "network_chain_id": "0x4",
      "network_code": "ETH",
      "network_display_name": "ETHEREUM",
      "network_logo_url": "",
      "rpc_url": "",
    }
  }
  setAddress={[MockFunction]}
  setCurrentChain={[MockFunction]}
  setModalOpen={[MockFunction]}
  setRemoveWalletListener={[MockFunction]}
  setSelectedWallet={[MockFunction]}
  walletList={
    Object {
      "Binance": Object {
        "id": "",
        "logo_url": "",
        "name": "Binance",
        "type": "",
      },
      "Metamask": Object {
        "id": "",
        "logo_url": "",
        "name": "Metamask",
        "type": "",
      },
      "Phantom": Object {
        "id": "",
        "logo_url": "",
        "name": "Phantom",
        "type": "",
      },
      "QRCode": Object {
        "id": "",
        "logo_url": "",
        "name": "QRCode",
        "type": "",
      },
      "WalletConnect": Object {
        "id": "",
        "logo_url": "",
        "name": "WalletConnect",
        "type": "",
      },
    }
  }
>
  <div>
    <div
      className="zamp-wallet-container"
      key="nikhin"
      onClick={[Function]}
      role="presentation"
    >
      <div
        className="zamp-wallet-label-container"
      >
        <img
          alt=""
          className="wallet-icons"
          src="undefined/public/icons/generic-crypto.svg"
        />
        <div
          className="wallet-name"
        >
          Pay with Exchange
        </div>
      </div>
      <div
        className="arrow-container"
      >
        <img
          alt=""
          height="15"
          src="undefined/public/icons/exchange-icons.svg"
          style="margin-right: 10px;"
        />
        <img
          alt=""
          height="11"
          src="undefined/public/images/right-arrow.svg"
        />
      </div>
    </div>
    <div
      className="zamp-wallet-container"
      key="nik"
      onClick={[Function]}
      role="presentation"
    >
      <div
        className="zamp-wallet-label-container"
      >
        <img
          alt=""
          className="wallet-icons"
          src="undefined/public/icons/pay-with-qr.svg"
        />
        <div
          className="wallet-name"
        >
          Pay with QR
        </div>
      </div>
      <div
        className="arrow-container"
      >
        <img
          alt=""
          height="11"
          src="undefined/public/images/right-arrow.svg"
        />
      </div>
    </div>
    <div
      className="zamp-wallet-container"
      key="Metamask"
      onClick={[Function]}
      role="presentation"
    >
      <div
        className="zamp-wallet-label-container"
      >
        <img
          alt="wallet-icon"
          className="wallet-icons"
          onError={[Function]}
          src=""
        />
        <div
          className="wallet-name"
        >
          Metamask
        </div>
      </div>
      <div
        className="arrow-container"
      >
        <img
          alt=""
          height="11"
          src="undefined/public/images/right-arrow.svg"
        />
      </div>
    </div>
  </div>
</WalletList>
`;
